### PR TITLE
feature: migrate to required components

### PR DESCRIPTION
**Pull request by patrickariel**
Closes https://github.com/StarArawn/bevy_ecs_tilemap/issues/597

# Objective

- Required components replaces bundles, but the crate still contains many user-facing bundles.

## Solution

- Deprecate the following bundles:
  - `TileBundle`
  - `MaterialTilemapBundle`
  - `TilemapBundle`
  - `StandardTilemapBundle`

- Migrate all examples to required components.
- Deprecate `MaterialTilemapHandle` in favor of `TilemapMaterial`. This is a pure name change done for two reasons:
  - To conform to the idiomatic pattern used [upstream](https://github.com/bevyengine/bevy/blob/71a9142deda547a872da3cc2b6a6300bad6e86a6/examples/3d/3d_shapes.rs#L83-L93):
    ```rust
    commands.spawn((
        Mesh3d(shape),
        MeshMaterial3d(material),
    ));
    ```
  - The material component is now something that is explicitly inserted by the user. So, we need to make sure that it's not cumbersome to remember and type.

## Testing

I've ran all the tests on Linux (Wayland) and they all look fine to me. But I'm not 100% familiar with all the quirks and edge-cases of this crate. So if I've overlooked any potential regressions, do let me know 🙂

## Migration Guide

The bundles are now all deprecated. With the new version, you should instead compose together `Tile`, `Tilemap` and `TilemapMaterial` to achieve your desired behavior.

Before:

```rust
commands.spawn(TileBundle::default());

commands.spawn(MaterialTilemapBundle {
    material,
    ..default(),
});

commands.spawn(TilemapBundle::default());

commands.spawn(StandardTilemapBundle::default());
```

After:

```rust
commands.spawn(Tile);

commands.spawn((Tilemap, TilemapMaterial(material)));

commands.spawn((Tilemap, TilemapMaterial::standard()));

commands.spawn(Tilemap);
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -112,11 +112,22 @@ impl Default for FrustumCulling {
...
 #[cfg(feature = "render")]
+#[expect(deprecated)]
+#[deprecated(
+    since = "0.16.0",
+    note = "Use the `Tilemap` and `TilemapMaterial::standard()` components instead. 
+        Inserting `Tilemap` will now also insert all necessary components automatically."
+)]
 pub type TilemapBundle = MaterialTilemapBundle<StandardTilemapMaterial>;
 
 #[cfg(feature = "render")]
```
Comment (ConnerPetzold) [inline src/lib.rs:123]: I think to maintain feature parity, you'd need to create two versions of the `Tilemap` component based on the render feature.

File: src/lib.rs [hunk 2]
```diff
@@ -136,14 +147,37 @@ pub struct MaterialTilemapBundle<M: MaterialTilemap> {
...
+    TilemapSpacing,
+    TileStorage,
+    TilemapTexture,
+    TilemapTileSize,
+    Transform,
+    TilemapRenderSettings,
+    TilemapAnchor,
+    Visibility,
+    FrustumCulling,
+    SyncToRenderWorld
```
Comment (ConnerPetzold) [inline src/lib.rs:169]: I wonder if `Tilemap` should require `TilemapMaterial<StandardTilemapMaterial>`? I'm not sure how generics work with required components but it seems like the vast majority of users won't need a custom material.

File: src/lib.rs (hunk 1 above)
Comment (patrickariel) [reply src/lib.rs:123]: I believe the feature split shouldn't be necessary anymore. `StandardTilemapBundle` (`not(feature = "render")`)  was just `MaterialTilemapBundle` (`feature = "render"`) without `TilemapMaterial`.

File: src/lib.rs (hunk 2 above)
Comment (patrickariel) [reply src/lib.rs:169]: I think this would sort of complicate the API surface, because now we would need a `TilemapWithoutMaterial` for users who wants custom material or no material (e.g. `not(feature = "render")`).

Having the material component be opt-in feels a lot more natural and idiomatic, in line with how it is handled upstream:

```rust
commands.spawn((
    Mesh3d(shape),
    MeshMaterial3d(material),
));
```

Here, `Mesh3d` does not automatically require `MeshMaterial3d<M>`. It must be inserted manually by the user.

Comment (ConnerPetzold) [reply src/lib.rs:169]: Hmm yeah I was only thinking about the use case of a custom material and not about the non-rendering case. I suppose that case will be more common for headless/backend.

Comment (ConnerPetzold) [reply src/lib.rs:169]: Although why not just make `Tilemap` conditionally require `TilemapMaterial<StandardTilemapMaterial>` based on the render flag to mirror what happens now? That way you don't need two separate components. And I think it still covers the use case of a custom material because I believe you'd still be able to override with a custom `TilemapMaterial`, though I'm not 100% sure about that.

Comment (teohhanhui) [reply src/lib.rs:169]: > I believe you'd still be able to override with a custom `TilemapMaterial`

That's correct.

Comment (patrickariel) [reply src/lib.rs:169]: > I believe you'd still be able to override with a custom `TilemapMaterial`, though I'm not 100% sure about that.

Unfortunately, this will not work. `TilemapMaterial<A>` and `TilemapMaterial<B>` are different types and will be treated as different components.

If you have a component like this:

```rust
#[derive(Component)]
#[require(TilemapMaterial<StandardTilemapMaterial>)]
struct Tilemap;
```

And you insert it like this:

```rust
commands.spawn((Tilemap, TileMapMaterial::<Custom>::default()));
```

You will end up with an entity that looks like this:

```rust
(
    Tilemap,
    TileMapMaterial<Custom>,
    TilemapMaterial<StandardTilemapMaterial>,
)
```

Comment (patrickariel) [reply src/lib.rs:169]: > We should follow the same pattern, i.e. the user needs to spawn `TilemapMaterial` by themselves when spawning `Tilemap`?

Yep, that's what I'm currently proposing.

Comment (ConnerPetzold) [reply src/lib.rs:169]: Ok here's my proposal:

- `Tilemap` conditionally requires `TilemapTexture` based on the `render` feature
- Make `TilemapMaterial` an optional component and remove `TilemapMaterial<StandardTilemapMaterial>` entirely
- If the entity has a `TilemapMaterial` component, use that as the material, otherwise fall back to using the `StandardTilemapMaterial` resource

I think that would allow:

```rust
(Tilemap) // Use the standard material
(Tilemap, TilemapMaterial<Custom>) // Use a custom material
```

Comment (ConnerPetzold) [reply src/lib.rs:169]: Ah yeah I see, that's a tricky problem..  What about something like this:
```rust
#[cfg(feature = "render")]
#[derive(Component, Debug, Default, Clone)]
#[require(
    TilemapGridSize,
    TilemapType,
    TilemapSize,
    TilemapSpacing,
    TileStorage,
    TilemapTexture,
    TilemapTileSize,
    Transform,
    TilemapRenderSettings,
    TilemapAnchor,
    Visibility,
    FrustumCulling,
    SyncToRenderWorld
)]
pub struct TilemapMaterial<M: MaterialTilemap>(pub Handle<M>);

#[cfg(feature = "render")]
pub type Tilemap = MaterialTilemapBundle<StandardTilemapMaterial>;

#[cfg(not(feature = "render"))]
#[derive(Component, Debug, Default, Clone)]
#[require(
    TilemapGridSize,
    TilemapType,
    TilemapSize,
    TilemapSpacing,
    TileStorage,
    TilemapTileSize,
    Transform,
    Visibility,
)]
pub struct Tilemap;
```

Comment (teohhanhui) [reply src/lib.rs:169]: @ConnerPetzold That just doesn't make sense and is super confusing for users. Why not just stick to the same pattern as Bevy? As a user I'd like better DX too, but I think this should be done at the Bevy level.

Comment (ConnerPetzold) [reply src/lib.rs:169]: I don't understand why it's confusing for users:
```rust
(TIlemap) // use the standard material if render=true, otherwise no materials/rendering
(TilemapMaterial<Custom>) // use a custom material, and will only be available if render=true
```

It's mirroring what ecs_tilemap has now, and essentially just collapses what you have with `Tilemap` and `TilemapMaterial`. I don't think the `MeshMaterial3d` pattern applies here because writing a custom tilemap material will be very rare. I think the more appropriate Bevy pattern to follow here would be `Sprite`, since like tilemaps you almost never need a custom material (and I guess in the case of Sprite, you can't even use a custom material). For the rare cases you do, you have the mesh/material escape hatch.

Comment (teohhanhui) [reply src/lib.rs:169]: I guess that would make sense if it's called `TilemapWithMaterial`.

**Comment by suprohub** [comment]
any updates ?
